src/main/java/org/quiltmc/loader/api/config/NightConfigSerializer.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.loader.api.config;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.electronwill.nightconfig.core.CommentedConfig;
+import com.electronwill.nightconfig.core.InMemoryCommentedFormat;
+import com.electronwill.nightconfig.core.UnmodifiableCommentedConfig;
+import com.electronwill.nightconfig.core.io.ConfigParser;
+import com.electronwill.nightconfig.core.io.ConfigWriter;
+import org.quiltmc.loader.api.config.annotations.Comment;
+import org.quiltmc.loader.api.config.values.CompoundConfigValue;
+import org.quiltmc.loader.api.config.values.ValueList;
+import org.quiltmc.loader.api.config.values.ValueMap;
+import org.quiltmc.loader.api.config.values.ValueTreeNode;
+
+public final class NightConfigSerializer<C extends CommentedConfig> implements Serializer {
+	private final String fileExtension;
+	private final ConfigParser<C> parser;
+	private final ConfigWriter writer;
+
+	public NightConfigSerializer(String fileExtension, ConfigParser<C> parser, ConfigWriter writer) {
+		this.fileExtension = fileExtension;
+		this.parser = parser;
+		this.writer = writer;
+	}
+
+	@Override
+	public String getFileExtension() {
+		return this.fileExtension;
+	}
+
+	@Override
+	public void serialize(Config config, OutputStream to) {
+		this.writer.write(write(createCommentedConfig(), config.nodes()), to);
+	}
+
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	@Override
+	public void deserialize(Config config, InputStream from) {
+		CommentedConfig read = this.parser.parse(from);
+
+		for (TrackedValue<?> trackedValue : config.values()) {
+			if (read.contains(trackedValue.getKey().toString())) {
+				((TrackedValue) trackedValue).setValue(coerce(read.get(trackedValue.getKey().toString()), trackedValue.getDefaultValue()), false);
+			}
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private static Object coerce(Object object, Object to) {
```

Comment:
Marshalling is a big can o' worms. This might want to be wholly set aside as its own class. It's kind of sad that java doesn't have builtin rules for it.

After marshalling to Classes and target objects, I've been moving very definitely towards marshalling to Types. You can extract a non-erased Type from a Field, and at that point you can follow the data hierarchy and stay in non-erased Types very easily. This also prevents you from having to construct default values for arbitrarily complex types.

Reply:
Could you perhaps give an example of marshalling using Types?